src/mpid/ch4/shm/xpmem/xpmem_avl.h

```diff
@@ -0,0 +1,564 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+#ifndef XPMEM_AVL_H_INCLUDED
+#define XPMEM_AVL_H_INCLUDED
+
+#include "mpidimpl.h"
+#include "xpmem_pre.h"
+#include <xpmem.h>
+/*
+=== BEGIN_MPI_T_CVAR_INFO_BLOCK ===
+
+cvars:
+    - name        : MPIR_CVAR_CH4_XPMEM_AVL_TREE_BUFFER_NUM
+      category    : CH4
+      type        : int
+      default     : 64
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        Maximal number of AVL tree node buffer pool we can have.
+
+      - name      : MPIR_CVAR_CH4_XPMEM_AVL_TREE_BUFFER_NODE_NUM
+      category    : CH4
+      type        : int
+      default     : 4096
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        The number of node each node buffer pool is able to provide.
+
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
+#define MPIDI_XPMEM_AVL_TREE_RLOW_LOCATE_MIN_RIGHT(node, rlow) (rlow >= node->min)
+#define MPIDI_XPMEM_AVL_TREE_RHIGH_LOCATE_MAX_LEFT(node, rhigh) (rhigh <= node->max)
+#define MPIDI_XPMEM_AVL_TREE_RLOW_LOCATE_LOW_LEFT(node, rlow) (rlow < node->low)
+#define MPIDI_XPMEM_AVL_TREE_RHIGH_LOCATE_HIGH_LEFT(node, rhigh) (rhigh <= node->high)
+
+/*
+    XPMEM AVL Tree segment caching contains four important operations:
+        - MPIDI_XPMEM_seg_avl_tree_init
+        - MPIDI_XPMEM_seg_avl_tree_search
+        - MPIDI_XPMEM_seg_avl_tree_insert
+        - MPIDI_XPMEM_seg_avl_tree_delete_all
+
+    MPIDI_XPMEM_seg_avl_tree_init:
+        This function initializes the basic attributes of AVL tree
+        including root node, tree size, pool count, node pool, and it
+        should be called only once for a AVL tree at the beginning of
+        the program.
+
+    MPIDI_XPMEM_seg_avl_tree_search:
+        This function performs binary search to find matched segment.
+        Assume we have an input segment from user, this function will
+        find a matched segment if the input segment low address and high
+        address locate between the recorded segment in AVL tree. If low
+        address of input segment is smaller than recorded segment, we
+        need to search left branch; or else, we need to search right
+        branch.
+
```

Comment:
What happens if the searching range is overlapped with two segments ? E.g., seg0[0x1000 - 0x2000], seg1[0x2000, 0x3000], searching range=[0x1500 - 0x2500]. In this case, do we insert a new segment[0x1500 - 0x2500] ?  Does `xpmem_attach` allow user to map the same physical range multiple times ?

Reply:
You need to call `xpmem_attach` to insert new segment for range=[0x1500 - 0x2500] because seg0 and seg1 in the attaching process could be noncontinuous. `xpmem_attach` indeed allows users to map the same range multiple times

Reply:
OK. So suppose segment0=[0x1000-0x2000], segment1[0x1000-0x3000], when I search [0x1000-0x2000], does it return whichever was cached first (i.e., stored in a higher node in the tree) ?

Reply:
Correct